pilot/pkg/serviceregistry/kube/controller/ambient/helpers.go

```diff
@@ -65,21 +67,59 @@ func serviceToAddressInfo(s *workloadapi.Service) model.AddressInfo {
 	}
 }
 
-func byteIPToString(b []byte) string {
-	ip, _ := netip.AddrFromSlice(b)
+func mustByteIPToString(b []byte) string {
+	ip, _ := netip.AddrFromSlice(b) // Address only comes from objects we create, so it must be valid
 	return ip.String()
 }
 
 func byteIPToAddr(b []byte) netip.Addr {
-	ip, _ := netip.AddrFromSlice(b)
+	ip, _ := netip.AddrFromSlice(b) // Address only comes from objects we create, so it must be valid
 	return ip
 }
 
-func (a *index) toNetworkAddress(vip string) *workloadapi.NetworkAddress {
+func (a *index) toNetworkAddress(vip string) (*workloadapi.NetworkAddress, error) {
+	ip, err := netip.ParseAddr(vip)
+	if err != nil {
+		return nil, fmt.Errorf("parse %v: %v", vip, err)
+	}
+	return &workloadapi.NetworkAddress{
+		Network: a.Network(vip, make(labels.Instance, 0)).String(),
+		Address: ip.AsSlice(),
+	}, nil
+}
+
+func (a *index) toNetworkAddressFromIP(ip netip.Addr) *workloadapi.NetworkAddress {
+	return &workloadapi.NetworkAddress{
+		Network: a.Network(ip.String(), make(labels.Instance, 0)).String(),
+		Address: ip.AsSlice(),
+	}
+}
+
+func (a *index) toNetworkAddressFromCidr(vip string) (*workloadapi.NetworkAddress, error) {
+	ip, err := parseCidrOrIP(vip)
+	if err != nil {
+		return nil, err
+	}
 	return &workloadapi.NetworkAddress{
 		Network: a.Network(vip, make(labels.Instance, 0)).String(),
-		Address: netip.MustParseAddr(vip).AsSlice(),
+		Address: ip.AsSlice(),
+	}, nil
+}
+
+// parseCidrOrIP parses an IP or a CIDR of a exactly 1 IP (e.g. /32).
+// This is to support ServiceEntry which supports CIDRs, but we don't currently support more than 1 IP
+func parseCidrOrIP(ip string) (netip.Addr, error) {
+	if strings.Contains(ip, "/") {
```

Comment:
```
	if strings.Contains(ip, "/") {
```

looks like `ParsePrefix` also does this same string check, so maybe just rely on that throwing a parse error?

Reply:
we might get slightly better error from this approach since we would otherwise pass an invalid CIDR to parseaddr. But probably pretty much the same